[Cocoa] WebCodecs H264 decoder should reorder frames according presentation time

youennf opened the pull request:
#### cf345a7967b324872fb7ff4a173307ddb93101ac
<pre>
[Cocoa] WebCodecs H264 decoder should reorder frames according presentation time
<a href="https://bugs.webkit.org/show_bug.cgi?id=263901">https://bugs.webkit.org/show_bug.cgi?id=263901</a>
<a href="https://rdar.apple.com/117692801">rdar://117692801</a>

Reviewed by Jean-Yves Avenard.

Add a SPS VUI parser to be able to extract max_num_reorder_frames.
Add an AVC parser to locate the SPS and parse it.

Compute the reorder queue size based on the parsed SPS.
Add a queue in RTCVideoDecoderH264 and reorder based on timestamp.

We will need a follow-up refactoring to pipe memory attribution of these video frames when getting them.

* LayoutTests/http/tests/webcodecs/h264-reordering-expected.txt: Added.
* LayoutTests/http/tests/webcodecs/h264-reordering.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
(RTCFrameDecodeParams::RTCFrameDecodeParams):
(RTCVideoFrameWithOrder::RTCVideoFrameWithOrder):
(RTCVideoFrameWithOrder::~RTCVideoFrameWithOrder):
(RTCVideoFrameWithOrder::take):
(decompressionOutputCallback):
(-[RTCVideoDecoderH264 init]):
(-[RTCVideoDecoderH264 decodeData:size:timeStamp:]):
(-[RTCVideoDecoderH264 setAVCFormat:size:width:height:]):
(-[RTCVideoDecoderH264 flush]):
(-[RTCVideoDecoderH264 processFrame:reorderSize:]):
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH265.mm:
(h265DecompressionOutputCallback):
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/nalu_rewriter.cc:
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/nalu_rewriter.h:

Canonical link: <a href="https://commits.webkit.org/271004@main">https://commits.webkit.org/271004@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/40ba642931d51aaa6e7832be5948238cead47022

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26706 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5320 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27946 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28918 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24417 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7145 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2717 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24329 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26968 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4136 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22927 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3636 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3675 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23919 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29403 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24350 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24319 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/29945 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3713 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1909 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27851 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5162 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/23655 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6499 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4182 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4066 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/nalu_rewriter.cc:
```diff
@@ -467,6 +468,236 @@ CMVideoFormatDescriptionRef CreateVideoFormatDescription(
+  max_dpb_frames = std::min(max_dpb_frames, 16ull);
+
+  if (spsAndVui->bitstream_restriction_flag) {
+    return std::min(spsAndVui->max_num_reorder_frames, max_dpb_frames);
```
Per spec
`max_num_ref_frames specifies the maximum number of short-term and long-term reference frames, complementary reference field pairs, and non-paired reference fields that may be used by the decoding     process for inter prediction of any picture in the sequence. max_num_ref_frames also determines the size of the sliding window operation as specified in subclause 8.2.5.3. The value of max_num_ref_frames shall be in the range of 0 to MaxDpbFrames (as specified in subclause A.3.1 or A.3.2), inclusive.)`

max_num_ref_frames defines the sliding window size, which is what defines the max depth of the needed re-ordering queue.

youennf replied:
According spec, sliding window size is internal to the VTB decoder. max_num_reorder_frames is for the output queue, which has to be  handled by VTB clients themselves. I think we align with the spec by using max_num_reorder_frames, which seems safer.

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/nalu_rewriter.cc:
```diff
@@ -467,6 +468,236 @@ CMVideoFormatDescriptionRef CreateVideoFormatDescription(
+    return std::min(spsAndVui->max_num_reorder_frames, max_dpb_frames);
+  }
+  if (spsAndVui->constraint_set3_flag && (spsAndVui->profile_idc == 44 || spsAndVui->profile_idc == 86 || spsAndVui->profile_idc == 100 || spsAndVui->profile_idc == 110 || spsAndVui->profile_idc == 122 || spsAndVui->profile_idc == 244)) {
+    return 0;
```
so this is what Chrome does, but I don't see why this would impact the re-ordering queue size:
like profile 44 is 2C which is `C44I` `CAVLC 4:4:4 Intra profile` 

Also, for reading max_num_reorder_frames it is stated:
`When the max_num_reorder_frames syntax element is not present, the value of max_num_reorder_frames value shall be inferred as follows: If profile_idc is equal to 44, 86, 100, 110, 122, or 244 and constraint_set3_flag is equal to 1, the value of max_num_reorder_frames shall be inferred to be equal to 0.`

This isn't what is being done here, it uses the profile_idc value and override the value even when max_num_reorder_frames is defined. 

In any case, `max_num_ref_frames` is the value that should be used, this is what you configure with the encoder . And the only party I see using max_num_reorder_frames is google weirdly enough.

youennf replied:
The spec defines this rule in different places. Each intra profile is clear about this though so this seems ok, see appendix A.2.9 for instance.

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
```diff
@@ -87,13 +116,16 @@ @implementation RTCVideoDecoderH264 {
   OSStatus _error;
   bool _useAVC;
+  std::deque<std::unique_ptr<RTCVideoFrameWithOrder>> _reorderQueue;
+  uint64_t _reorderSize;
```
reorder size can't be higher than 16 frames, so using a uint64_t is wasting space a bit. a uint8_t would do

youennf replied:
ok, I'll change this.

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
```diff
@@ -378,4 +417,20 @@ - (NSString *)implementationName {
+      return a->timeStamp < b->timeStamp;
+    });
+    while (_reorderQueue.size() > _reorderQueue.front()->reorderSize) {
+      auto* frame = _reorderQueue.front()->take();
```
` auto *` ? for obj_c style?

youennf replied:
ok

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
```diff
@@ -76,7 +105,7 @@ void decompressionOutputCallback(void *decoderRef,
                                 timeStampNs:CMTimeGetSeconds(timestamp) * rtc::kNumNanosecsPerSec];
   decodedFrame.timeStamp = decodeParams->timestamp;
-  decodeParams->callback(decodedFrame);
+  [decoder processFrame:decodedFrame reorderSize:decodeParams->reorderSize];
```
if I understand `decompressionOutputCallback` is called on the encoder thread, but flush would be on a caller thread. It could mean the queue is accessed from 2 threads at once

youennf replied:
flush is calling VTDecompressionSessionWaitForAsynchronousFrames(), so it will block until all frames are given to `decompressionOutputCallback`. Once VTDecompressionSessionWaitForAsynchronousFrames() is completed, `decompressionOutputCallback` will not be called until we call again decode. So we can safely call _callback in flush().

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
```diff
@@ -87,13 +116,16 @@ @implementation RTCVideoDecoderH264 {
   RTCVideoDecoderCallback _callback;
   OSStatus _error;
   bool _useAVC;
+  std::deque<std::unique_ptr<RTCVideoFrameWithOrder>> _reorderQueue;
```
for safety, I would define on which thread _reorderQueue can be accessed (through `WTF_GUARDED_BY_CAPABILITY`) because at it looks, it seems to be racy

youennf replied:
We cannot really do that in flush. I guess we could introduce a queue for that specific purpose so that the decoder client could tell where to call callback.

jyavenard replied:
isn't there a contract on which thread callback will be called on ?

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
```diff
@@ -359,6 +394,10 @@ - (void)destroyDecompressionSession {
   if (_decompressionSession)
     VTDecompressionSessionWaitForAsynchronousFrames(_decompressionSession);
+  while (!_reorderQueue.empty()) {
+    _callback(_reorderQueue.front()->take());
```
so here _callback is called on the caller's queue.

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
```diff
@@ -378,4 +417,20 @@ - (NSString *)implementationName {
+    }
+    return;
+  }
+  _callback(decodedFrame);
```
and here _callback is called on the encoder's queue

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
```diff
@@ -87,13 +117,17 @@ @implementation RTCVideoDecoderH264 {
   RTCVideoDecoderCallback _callback;
   OSStatus _error;
   bool _useAVC;
+  std::deque<std::unique_ptr<RTCVideoFrameWithOrder>> _reorderQueue;
```
do WebRTC::lock work with `WTF_GUARDED_BY_LOCK(_reorderQueueLock)` it would be better to have that there.
Or at least a comment that it must be accessed with _reorderQueueLock held.

Personally, I like to put the lock definition above the variables it protects, in a paragraph-like style

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
```diff
@@ -49,6 +53,32 @@ static void overrideColorSpaceAttachments(CVImageBufferRef imageBuffer) {
   CVBufferSetAttachment(imageBuffer, (CFStringRef)@"ColorInfoGuessedBy", (CFStringRef)@"RTCVideoDecoderH264", kCVAttachmentMode_ShouldPropagate);
 }
 
+struct RTCVideoFrameWithOrder {
```
do we need this object when RTCVideoFrame already has a timeStamp attribute?

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
```diff
@@ -49,6 +53,32 @@ static void overrideColorSpaceAttachments(CVImageBufferRef imageBuffer) {
+    }
+
+    void* frame;
+    uint64_t timeStamp;
```
`const`

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
```diff
@@ -49,6 +53,32 @@ static void overrideColorSpaceAttachments(CVImageBufferRef imageBuffer) {
+
+    void* frame;
+    uint64_t timeStamp;
+    uint64_t reorderSize;
```
`const`

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
```diff
@@ -156,6 +190,7 @@ - (NSInteger)decodeData:(const uint8_t *)data
   rtc::ScopedCFTypeRef<CMVideoFormatDescriptionRef> inputFormat =
       rtc::ScopedCF(webrtc::CreateVideoFormatDescription(data, size));
   if (inputFormat) {
+    _reorderSize = webrtc::ComputeH264ReorderSizeFromAnnexB(data, size);
```
I think we should check that the first packet we received contains a SPS/PPS NAL rather than assume 0 (you'll get a decode error anyway)

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
```diff
@@ -359,6 +396,12 @@ - (void)destroyDecompressionSession {
+
+  webrtc::MutexLock lock(&_reorderQueueLock);
+  while (!_reorderQueue.empty()) {
+    _callback(_reorderQueue.front()->take());
```
is there a chance that the callback insert a new frame which would cause a deadlock?

jyavenard commented on Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
```diff
@@ -378,4 +421,21 @@ - (NSString *)implementationName {
+    });
+    while (_reorderQueue.size() > _reorderQueue.front()->reorderSize) {
+      auto *frame = _reorderQueue.front()->take();
+      _callback(frame);
```
same here. what if the callback add a new frame?

jyavenard reviewed (approved):
r=me with nit

youennf committed:
[Cocoa] WebCodecs H264 decoder should reorder frames according presentation time
https://bugs.webkit.org/show_bug.cgi?id=263901
rdar://117692801

Reviewed by Jean-Yves Avenard.

Add a SPS VUI parser to be able to extract max_num_reorder_frames.
Add an AVC parser to locate the SPS and parse it.

Compute the reorder queue size based on the parsed SPS.
Add a queue in RTCVideoDecoderH264 and reorder based on timestamp.

We will need a follow-up refactoring to pipe memory attribution of these video frames when getting them.

* LayoutTests/http/tests/webcodecs/h264-reordering-expected.txt: Added.
* LayoutTests/http/tests/webcodecs/h264-reordering.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH264.mm:
(RTCFrameDecodeParams::RTCFrameDecodeParams):
(RTCVideoFrameWithOrder::RTCVideoFrameWithOrder):
(RTCVideoFrameWithOrder::~RTCVideoFrameWithOrder):
(RTCVideoFrameWithOrder::take):
(decompressionOutputCallback):
(-[RTCVideoDecoderH264 init]):
(-[RTCVideoDecoderH264 decodeData:size:timeStamp:]):
(-[RTCVideoDecoderH264 setAVCFormat:size:width:height:]):
(-[RTCVideoDecoderH264 flush]):
(-[RTCVideoDecoderH264 processFrame:reorderSize:]):
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/RTCVideoDecoderH265.mm:
(h265DecompressionOutputCallback):
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/nalu_rewriter.cc:
* Source/ThirdParty/libwebrtc/Source/webrtc/sdk/objc/components/video_codec/nalu_rewriter.h:

Canonical link: https://commits.webkit.org/271004@main
